### PR TITLE
Introduce Cancellable API

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

See #2128 

All the time Gradle can't kill the build running in daemon. For example, if we try to press Ctrl-C, Gradle will mark the daemon as `Cancelled` then wait 10s for the build to finish (do nothing but wait). Unfortunately in most cases 10s is not enough, so you will see a few lines in daemon log:

```
Cancel: daemon is still busy after grace period. Will force stop.
...
Daemon vm is shutting down...
```

The consequence is, in most cases, pressing Ctrl-C will kill the daemon.

This PR introduces an interface `Cancellable` to allow tasks to react to the Ctrl-C. Cancellable tasks' `cancel()` callbacks will be executed upon cancellation.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.execution;
+
+/**
+ * Indicates that a task is cancellable.
+ */
+public interface Cancellable {
```
Comment (ghale) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: If we're going to call this `Cancellable` then I suggest we remove references to tasks in the comments as it implies it represents anything that could be cancellable (and we might want potentially want to use it for other cancellable things).

Or is this intended to _only_ represent cancellable tasks?  In which case we should probably call it `CancellableTask`.

File: subprojects/core/src/main/java/org/gradle/process/internal/JavaExecAction.java [hunk 2]
```diff
@@ -22,4 +22,6 @@
 @NonExtensible
 public interface JavaExecAction extends JavaExecSpec {
     ExecResult execute();
+
+    void cancel();
```
Comment (ghale) [inline subprojects/core/src/main/java/org/gradle/process/internal/JavaExecAction.java]: Could this just implement `Cancellable`?  Meaning "this is a thing that can be cancelled when the build is cancelled".

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java (hunk 1 above)
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: Agreed, this should be task-agnostic.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * When Ctrl-C is pressed during the task execution, this method of that task will be invoked.
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: ❌ Builds can be cancelled in other ways. This should not get into the detail how the build was cancelled.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        if (task instanceof Cancellable) {
+            Runnable onCancel = new Runnable() {
+                @Override
+                public void run() {
+                    Cancellable.class.cast(task).cancel();
+                }
+            };
+            buildCancellationToken.addCallback(onCancel);
+            delegate.execute(task, state, context);
+            buildCancellationToken.removeCallback(onCancel);
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java]: ❌ Should be in a finally-block

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    private final TaskExecuter delegate;
+    private final BuildCancellationToken buildCancellationToken;
+
+    public CancellableTaskExecuter(TaskExecuter delegate, BuildCancellationToken buildCancellationToken) {
+        this.delegate = delegate;
+        this.buildCancellationToken = buildCancellationToken;
+    }
+
+    @Override
+    public void execute(final TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java]: ❌ Not sure if we have test cases for this already:

- cancelled tasks don't lead to weird errors on the console
- outputs of cancelled tasks are not snapshotted/cached, the task is rerun the next time

File: subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java [hunk 6]
```diff
@@ -57,7 +59,7 @@
  * gradle someJavaExecTask --debug-jvm
  * </pre>
  */
-public class JavaExec extends ConventionTask implements JavaExecSpec {
+public class JavaExec extends ConventionTask implements JavaExecSpec, Cancellable {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java]: ❌ I don't think we should do this at the task level, but instead make *every* exec action (no matter if through a task or invoked via `Project.exec`) cancellable.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ExecuteBuild.java [hunk 7]
```diff
@@ -56,14 +56,12 @@ protected void doBuild(final DaemonCommandExecution execution, Build build) {
...
-                    @Override
-                    public void run() {
-                        LOGGER.info(DaemonMessages.CANCELED_BUILD);
-                    }
-                });
-            }
+            buildRequestContext.getCancellationToken().addCallback(new Runnable() {
+                @Override
+                public void run() {
+                    LOGGER.info(DaemonMessages.CANCELED_BUILD);
```
Comment (oehme) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ExecuteBuild.java]: Why this change?

File: subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        taskIsCancellable()
+    }
+
+    void taskIsCancellable() {
+        def client = new DaemonClientFixture(executer.withArgument("--debug").withTasks("block").start())
+        waitFor(START_UP_MESSAGE)
+        daemons.daemon.assertBusy()
+
+        client.kill()
+        waitFor('BUILD FAILED in')
```
Comment (oehme) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy]: ❌ It should not be failed, but cancelled.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java (hunk 1 above)
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: After thinking about this some more, I think we should split this up into two PRs:

1. Make all exec actions cancellable. This doesn't require any new API, but we can already ensure that all the correct behavior is there (e.g. task is rerun the next time etc.)
2. Add a public API. Not sure if it should be an interface you implement or rather rather a callback-based service like `CancellationToken`, as that gives users more flexibility in how they structure their code.

File: subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java (hunk 6 above)
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java]: @ghale I'd be interested in your take on this - Would there be a reason to only do this for some exec actions or is a transparent cancellation of all of them reasonable?

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ExecuteBuild.java (hunk 7 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/ExecuteBuild.java]: Previously, only in continuous mode this message is printed because only continuous build can respond to cancellation. Now all builds can respond to cancellation so I think it's necessary to remove this limitation.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java (hunk 1 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: Good idea. After thinking a lot, I'm going to only implement `Cancellable` API in this PR. As the first step, this PR only contains a general task cancellation (not including `Exec` cancellation). I'm not going to give users "more flexibility" in this PR, for example, should this action be able to react to cancellation?

```
class MyAction implements Action, Cancellable {
     void execute(Task t){ ... }
     void cancel() { ... }
}

task t {
     doLast(new MyAction())
}
```

Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: Exec cancellations are supposed to be next PR.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java (hunk 4 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java]: Done. Thanks.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java (hunk 3 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java]: Done. Thanks.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java (hunk 5 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java]: Added the test cases here: https://github.com/gradle/gradle/blob/0211eb748345bc88222b985530c1383ddb479232/subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy#L64

Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java]: Missing test case for cachable tasks.

Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CancellableTaskExecuter.java]: Added test cases with `--build-cache` enabled and assertions of no build status being printed on client console.

File: subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java (hunk 6 above)
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/api/tasks/JavaExec.java]: I think we really have two different goals here.  One is that we want exec actions to be cancellable (wherever they occur) and two is that we want any arbitrary task to be cancellable.  I think the former covers tasks that exec things, and the latter covers exec actions that are a part of tasks.  Realistically, I don't think there is an order here - we can do either without necessarily doing the other.  But, per your other comments, we should split the story up into parts.  I also like the idea of a more generic cancellation api rather than just cancelling tasks.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                volatile Thread taskExecutionThread
+
+                @TaskAction
+                void run() {
+                    println '$START_UP_MESSAGE'
+                    taskExecutionThread = Thread.currentThread()
+                    try {
+                        new java.util.concurrent.CountDownLatch(1).await()
+                    } catch (InterruptedException e) {
+                        ${scenario == 'swallowed' ? '' : 'throw e'}
```
Comment (JLLeitschuh) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy]: You should always be reseting the thread's interrupted flag when you catch this exception.
`Thread.currentThread().interrupt()`

Comment (oehme) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy]: Sure but we can't trust Gradles correctness on people doing that.

Comment (JLLeitschuh) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy]: Ah, I guess I'll have to agree with you there.

**Comment by oehme** [comment]
Closing as this was superseded

**Comment by oehme** [comment]
https://github.com/gradle/gradle/pull/5249

A public API will require some more discussion first. Meanwhile the above fixes the most common problem.
